### PR TITLE
Managing organization permissions

### LINKED ISSUE #708: Managing organization permissions
### What should be implemented:
- [x] `ManageUserRoleCard` functional component
- [x] `LnkUserOrganizationController` which will support role changing on backend
- [x] `LnkUserOrganizationTest`
- [x] `OrganizationSettingsMenu` that will contain `ManageUserRoleCard`

`ManageUserRoleCard` should be able to:
- [x] Show all the users with their roles
- [x] Correctly show self role
- [x] Update users permissions
- [x] Support removing people from group
- [x] Support adding people to group
- [x] Autocomplete names of candidates to participate

### PR DESCRIPTION
**Pull request by sanyavertolet**
This PR closes #708

What's done:
- [x] Implemented ManageUserRoleCard functional component
- [x] Added LnkUserOrganizationController
- [x] Added LnkUserOrganizationTest
- [x] Added OrganizationSettingsMenu that contains ManageUserRoleCard
- [x] Added ManageUserRoleCard into the ProjectSettingsMenu
- [x] Added OrganizationStatus that marks organizations as created or deleted
- [x] Added better database data for local frontend testing

What is left out of this PR:
 * Support deleting project on backend
 * Support deleting organization on backend

ProjectSettingsMenu:
Снимок экрана 2022-05-04 в 12 58 56

OrganizationSettingsMenu:
Снимок экрана 2022-05-04 в 12 57 37

### CODE REVIEW

File: db/v-2/insert-test-data/sqlRequests/user.csv [hunk 1]
```diff
@@ -0,0 +1,13 @@
+id;name;password;role;source;avatar;email;company;location;linkedin;git_hub;twitter
+1;"admin";"{bcrypt}$2a$12$7jB6.fJ/uHZiti94w0.QOu0M/lPzKU1lvX9TLL3uZHFjLGIfz1gHq";"ROLE_SUPER_ADMIN";"basic";"null";"null";"null";"null";"null";"null";"null"
+2;"user";"noop{123}";"ROLE_VIEWER";"basic";"null";"null";"null";"null";"null";"null";"null"
```
Comment (Cheshiriks) [inline db/v-2/insert-test-data/sqlRequests/user.csv:3]: You don't need to add _null_ here. It is enough not to write anything between semicolons 
2;"user";"noop{123}";"ROLE_VIEWER";"basic";;;;;;

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /**
+     * @param organizationName
+     * @param userName
+     * @param authentication
+     * @return role of user with [userName] in organization with [organizationName]
+     * @throws ResponseStatusException
+     */
+    @GetMapping("/organizations/roles/{organizationName}")
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt:70]: Yes, that makes sense. The same should be done with projects.

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+     */
+    var groupPath: String
+
+    /**
+     * Kind of a group that will be shown ("project" or "organization" for now)
+     */
+    var groupType: String
+}
+
+private fun String.toRole() = when (this) {
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt:56]: That is a way better, thanks

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/LnkUserProjectService.kt [hunk 4]
```diff
@@ -63,7 +65,7 @@ class LnkUserProjectService(
             lnkUserProjectRepository.deleteById(it)
         }
         ?: run {
-            throw NoSuchElementException("Cannot delete user ${user.name ?: user.id} from project ${project.organization.name}/${project.name}: no such link was found.")
+            throw ResponseStatusException(HttpStatus.NOT_FOUND)
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/LnkUserProjectService.kt:68]: Ok, thank you. Will fix it.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        @RequestParam prefix: String,
+        authentication: Authentication,
+    ): List<UserInfo> {
+        if (prefix.isEmpty()) {
+            return emptyList()
+        }
+        val organization = organizationService.findByName(organizationName)
+            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+        val organizationUserIds = lnkUserOrganizationService.getAllUsersAndRolesByOrganization(organization)
+            .map { (users, _) ->
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt:187]: :)

File: save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     * @param organization
+     * @param permission
+     * @return whether user described by [authentication] can have [permission] on [organization]
+     */
+    fun hasPermission(authentication: Authentication?, organization: Organization, permission: Permission): Boolean {
+        authentication ?: return false
+        val userId = (authentication.details as AuthenticationDetails).id
+        val organizationRole = lnkUserOrganizationService.findRoleByUserIdAndOrganization(userId, organization)
+
+        if (authentication.hasRole(Role.SUPER_ADMIN)) {
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt:31]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ProjectController.kt [hunk 7]
```diff
@@ -189,6 +189,7 @@ class ProjectController(private val projectService: ProjectService,
     fun updateProject(@RequestBody project: Project, authentication: Authentication): Mono<StringResponse> = projectService.findWithPermissionByNameAndOrganization(
         authentication, project.name, project.organization, Permission.WRITE
     )
+        .filter { projectPermissionEvaluator.hasPermission(authentication, project, Permission.WRITE) }
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ProjectController.kt:192]: I have opened an issue to fix it (#723).

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/OrganizationService.kt [hunk 8]
```diff
@@ -69,14 +72,28 @@ class OrganizationService(
...
+        val otherRole = lnkUserOrganizationService.findRoleByUserIdAndOrganization(otherUser.id!!, organization)
+        return isOrganizationAdminOrHigher(selfRole) && hasAnotherUserLessPermissions(selfRole, otherRole) &&
+                isRequestedPermissionsCanBeSetByUser(selfRole, requestedRole)
     }
+
+    private fun hasAnotherUserLessPermissions(selfRole: Role, otherRole: Role): Boolean = selfRole.priority > otherRole.priority
+
+    private fun isRequestedPermissionsCanBeSetByUser(selfRole: Role, requestedRole: Role): Boolean = selfRole.priority > requestedRole.priority
+
+    private fun isOrganizationAdminOrHigher(userRole: Role): Boolean = userRole.priority >= Role.ADMIN.priority
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/OrganizationService.kt:98]: Yes, and the same should be done with `ProjectPermissionEvaluator`

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 9]
```diff
@@ -269,7 +269,7 @@ class ExecutionController(private val executionService: ExecutionService,
             IllegalArgumentException("Can't rerun execution $id, because it does not exist")
         }
         if (!projectPermissionEvaluator.hasPermission(
-            authentication, execution.project, Permission.WRITE
```
Comment (kgevorkyan) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:272]: Why did you change this?
As we allow to submit execution only with `write` permissions, probably `rerun` should have the same restriction

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+import kotlinx.html.js.onChangeFunction
+import kotlinx.html.js.onClickFunction
+import kotlinx.js.jso
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+/**
+ * [RProps] for card component
+ */
+external interface ManageUserRoleCardProps : PropsWithChildren {
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt:39]: Seems like no, thanks

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 9 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:272]: oh, changed by mistake

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+     */
+    @GetMapping("{organizationName}/users")
+    fun getAllUsersByOrganizationName(
+        @PathVariable organizationName: String,
+        authentication: Authentication,
+    ): List<UserInfo> {
+        val organization = organizationService.findByName(organizationName)
+            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+        val usersWithRoles = lnkUserOrganizationService.getAllUsersAndRolesByOrganization(organization)
+            .also { logger.debug(it.toString()) }
```
Comment (kgevorkyan) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt:53]: I think it will be hard to understand where did this message comes from, when we will look into debug logs, since there is no any additional words about what is printing

Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt:53]: Yes, it was only for debugging. Not sure if we need such logs.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+import org.cqfn.save.permission.Permission
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.security.core.Authentication
+import org.springframework.stereotype.Component
+
+/**
+ * Class that is capable of assessing user's permissions regarding organizations.
+ */
+@Component
+class OrganizationPermissionEvaluator {
```
Comment (kgevorkyan) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt:16]: Some logic looks quite similar to the `ProjectPermissionEvaluator`, could you please look, can we extract common parts into the separate module? Probably we can accept in these methods some lambdas as last arguments, which will contain     individual logic for project or organization, but the most part of permissions evaluation will be the same?

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+     */
+    var groupPath: String
+
+    /**
+     * Kind of a group that will be shown ("project" or "organization" for now)
+     */
+    var groupType: String
+}
+
+// private fun String.toRole() = when (this) {
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt:53]: commented code?

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+ * @return a functional component representing a role managing card
+ */
+@Suppress(
+    "LongMethod",
+    "EMPTY_BLOCK_STRUCTURE_ERROR",
+    "TOO_LONG_FUNCTION",
+    "MAGIC_NUMBER",
+    "ComplexMethod",
+)
+fun manageUserRoleCardComponent(
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt:80]: Any screenshots of the new components in PR description?

File: save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt (hunk 12 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt:16]: Well there are some common parts but they are kinda small. The point is that `ProjectPermissionEvaluator` is bigger and has more cases to check. To my mind it is better to split this logic

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt (hunk 13 above)
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt:53]: :)

File: save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt (hunk 12 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt:16]: There are some common parts of code, but to my mind they are too small to move these two permission evaluators to the same class. To my mind it is better to keep the logic split.

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt (hunk 14 above)
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ManageUserRoleCard.kt:80]: Ok, Uploaded to PR.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+     * @param authentication
+     * @return role of user with [userName] in organization with [organizationName]
+     * @throws ResponseStatusException
+     */
+    @GetMapping("/{organizationName}/users/roles")
+    @ApiResponse(responseCode = "200", description = "Successfully fetched user's role")
+    @ApiResponse(
+        responseCode = "404", description = "Requested user or organization doesn't exist."
+    )
+    fun getRole(
```
Comment (Cheshiriks) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt:73]: This is duplicate fun _getRoleOrganization_ in OrganizationController

**Review by Cheshiriks — CHANGES_REQUESTED** [review]

**Review by petertrr — APPROVED** [review]

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt (hunk 15 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/LnkUserOrganizationController.kt:73]: Ok, removed `getRoleOrganization` from `OrganizationController`

**Review by Cheshiriks — APPROVED** [review]

File: save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt (hunk 12 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/security/OrganizationPermissionEvaluator.kt:16]: Discussed with @petertrr , decided to leave as is because in the nearest future this classes will differ more.
